File: src/organize/metadata/acoustIdClient.ts
/**
 * TypeScript/`fetch` re-port of
 * `legacy-android-native-attempt/.../organize/metadata/AcoustIdClient.kt`.
 * Resolves a Chromaprint fingerprint (see `organize/fingerprint/` - fingerprint
 * *generation* is a separate, not-yet-implemented spike, see that module) to
 * MusicBrainz recording IDs via AcoustID, for the classical-music renamed-title
 * fallback in Claude/MUSIC ORGANIZATION.md. `apiKey` is the user's own free
 * AcoustID client key, from `organize/settings/secureSettings.ts` - never hardcoded.
 *
 * Note: AcoustID's API defaults to XML unless `format=json` is passed explicitly -
 * the Kotlin version omitted this (a latent bug never caught since nothing ever
 * actually called it against the live API); fixed here.
 */

export interface AcoustIdMatch {
  recordingId: string;
  title: string | null;
  artist: string | null;
  score: number;
}

interface AcoustIdRecording {
  id?: string;
  title?: string;
  artists?: Array<{ id?: string; name?: string }>;
}

interface AcoustIdResult {
  id?: string;
  score?: number;
  recordings?: AcoustIdRecording[];
}

interface AcoustIdLookupResponse {
  status?: string;
  results?: AcoustIdResult[];
}

/** Returns matches sorted best-first, or an empty list if no key is configured,
 * nothing matched, or the lookup failed. */
export async function lookupFingerprint(
  apiKey: string | null,
  fingerprint: string,
  durationSeconds: number,
): Promise<AcoustIdMatch[]> {
  if (!apiKey) return [];

  const params = new URLSearchParams({
    client: apiKey,
    duration: String(durationSeconds),
    fingerprint,
    meta: 'recordings',
    format: 'json',
  });

  let data: AcoustIdLookupResponse;
  try {
    const response = await fetch(`https://api.acoustid.org/v2/lookup?${params.toString()}`);
    if (!response.ok) return [];
    data = await response.json();
  } catch {
    return [];
  }

  return (data.results ?? [])
    .slice()
    .sort((a, b) => (b.score ?? 0) - (a.score ?? 0))
    .flatMap((result) =>
      (result.recordings ?? [])
        .filter((recording) => Boolean(recording.id))
        .map(
          (recording): AcoustIdMatch => ({
            recordingId: recording.id!,
            title: recording.title ?? null,
            artist: recording.artists?.map((a) => a.name ?? '').join(', ') || null,
            score: result.score ?? 0,
          }),
        ),
    );
}

File: app/src/main/java/com/mslynch/awesomesource/organize/metadata/AcoustIdClient.kt
package com.mslynch.awesomesource.organize.metadata

import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory

/**
 * Resolves a Chromaprint fingerprint (see
 * [com.mslynch.awesomesource.organize.fingerprint.Fingerprinter] - fingerprint
 * *generation* is a separate, not-yet-implemented spike) to MusicBrainz recording
 * IDs via AcoustID, for the classical-music renamed-title fallback in
 * Claude/MUSIC ORGANIZATION.md. [apiKey] is the user's own free AcoustID client key,
 * from [com.mslynch.awesomesource.organize.settings.SecureSettings] - never hardcoded.
 *
 * Note: the key currently on file is a *personal/user* AcoustID key (for submitting
 * fingerprints), not an *application* key (needed for this lookup's `client=`
 * parameter) - lookups will return "invalid API key" until it's replaced with one
 * from https://acoustid.org/new-applications (found during the Expo/TypeScript
 * attempt; unresolved regardless of stack).
 */
class AcoustIdClient(
    private val apiKey: String?,
    okHttpClient: OkHttpClient = OkHttpClient(),
) {
    data class Match(val recordingId: String, val title: String?, val artist: String?, val score: Double)

    private val api: AcoustIdApi = Retrofit.Builder()
        .baseUrl("https://api.acoustid.org/v2/")
        .client(okHttpClient)
        .addConverterFactory(MoshiConverterFactory.create())
        .build()
        .create(AcoustIdApi::class.java)

    /** Returns matches sorted best-first, or an empty list if no key is configured,
     * nothing matched, or the lookup failed. */
    suspend fun lookup(fingerprint: String, durationSeconds: Int): List<Match> {
        val key = apiKey ?: return emptyList()
        val response = runCatching { api.lookup(key, durationSeconds, fingerprint) }.getOrNull() ?: return emptyList()
        return response.results.orEmpty()
            .sortedByDescending { it.score ?: 0.0 }
            .flatMap { result ->
                result.recordings.orEmpty().map { recording ->
                    Match(
                        recordingId = recording.id.orEmpty(),
                        title = recording.title,
                        artist = recording.artists?.joinToString(", ") { it.name.orEmpty() },
                        score = result.score ?: 0.0,
                    )
                }
            }
            .filter { it.recordingId.isNotEmpty() }
    }
}
